Feat/add nfc cards

AnezkaQuanti opened the pull request.

AnezkaQuanti committed:
FEAT: basic repo classes for NFC cards

AnezkaQuanti committed:
FEAT: NFC added

AnezkaQuanti committed:
FIX: Nfc card error messages

AnezkaQuanti committed:
FIX: Voucher and Vendor entities changed

balakzde commented on src/vendor_app/app/src/main/kotlin/cz/quanti/android/vendor_app/main/checkout/viewmodel/CheckoutViewModel.kt:
```diff
@@ -66,13 +75,45 @@ class CheckoutViewModel(
         shoppingHolder.chosenCurrency = ""
     }
 
+    fun payByCard(pin: String, value: Double): Observable<UserBalance> {
```
Tady to vypadá, že by bylo vhodnější vracet Single

balakzde commented on src/vendor_app/app/src/main/kotlin/cz/quanti/android/vendor_app/utils/VendorUtil.kt:
```diff
@@ -30,3 +34,24 @@ fun isNetworkAvailable(context: Context): Boolean {
     return networkInfo != null && networkInfo.isConnected
 }
+
+fun getDbMigration1to2(): Migration {
```
tuhle metodu bych spíše viděl v VendorDb companion objectu

balakzde commented on src/vendor_app/app/src/main/kotlin/cz/quanti/android/vendor_app/main/checkout/fragment/ScanCardFragment.kt:
```diff
@@ -0,0 +1,184 @@
+                        )
+                    }, {
+                        when (it) {
+                            is PINException -> {
```
rozděl toto do více menších metod. Asi ten error handling by se dal dát klidně zvlášť

balakzde commented on src/vendor_app/app/src/main/kotlin/cz/quanti/android/vendor_app/main/checkout/fragment/ScanCardFragment.kt:
```diff
@@ -0,0 +1,184 @@
+        )
+
+        init()
+        payByCard()
```
Neprovolá se tohle jen jednou? Nejsem si tim vůbec jistej, ale skoro bych řek, že když třeba nemáš zapnutý NFC a hodí tě to do nastavení a pak se vrátíš zpět, tak už se onViewCreated nezavolá. (Ale jistej si nejsem, to jen aby sis to zkontrolovala)

AnezkaQuanti replied:
No to mi uplne nevadi. Ja v te metode otviram dialog, do ktereho se zadava pin, a ten tam zustane i pote, co se vrati uzivatel z nastaveni. Az po vyplneni pinu se spousti samotne skenovani :) (vyzkousela jsem si vypnout nfc a vse fungovalo jak by melo)

AnezkaQuanti committed:
change entities for new API

# Conflicts:
#	src/vendor_app/app/src/main/kotlin/cz/quanti/android/vendor_app/di/KoinInitializer.kt
#	src/vendor_app/app/src/main/kotlin/cz/quanti/android/vendor_app/main/checkout/viewmodel/CheckoutViewModel.kt
#	src/vendor_app/app/src/main/kotlin/cz/quanti/android/vendor_app/repository/VendorDb.kt
#	src/vendor_app/app/src/main/kotlin/cz/quanti/android/vendor_app/utils/VendorUtil.kt

AnezkaQuanti committed:
FEAT: Changed purchasing products with vouchers related data to match new API

AnezkaQuanti committed:
FIX: Total text rounding negative values fixed

AnezkaQuanti committed:
FEAT: Blocked smartcards basic logic added

AnezkaQuanti committed:
FEAT: Get rid of quantity

AnezkaQuanti committed:
FEAT: POST to PATCH on card purchase

AnezkaQuanti committed:
FEAT: Card currency

AnezkaQuanti committed:
FEAT: Database changed to match new API data for purchases

AnezkaQuanti committed:
REF: removed one TODO

AnezkaQuanti committed:
FIX: PIN-208 New Vendor App Multi Bug Report (login screen, app name and flat buttons)

AnezkaQuanti committed:
FIX: PIN-208 New Vendor App Multi Bug Report (shopping cart -> basket)

AnezkaQuanti committed:
FIX: PIN-208 New Vendor App Multi Bug Report (action bar changes)

AnezkaQuanti committed:
FIX: PIN-208 New Vendor App Multi Bug Report (default currency)

AnezkaQuanti committed:
FIX: PIN-208 New Vendor App Multi Bug Report (products and cart on one page and some arabic translation)

AnezkaQuanti committed:
FEAT: card pin restricted only to 4 numbers

AnezkaQuanti committed:
FIX: disable scanning of NFC when its not needed

AnezkaQuanti committed:
FIX: Error messages during card payment improved and success is shown in dialog instead of toast

AnezkaQuanti committed:
FIX: minor string stuff

AnezkaQuanti committed:
FIX: minor string stuff

balakzde commented on src/vendor_app/app/src/main/kotlin/cz/quanti/android/vendor_app/main/checkout/fragment/ScanCardFragment.kt:
```diff
@@ -0,0 +1,196 @@
+                                Toast.LENGTH_LONG
+                            ).show()
+                        }
+                        is VendorAppException -> {
```
tohle vypadá stejně jako ta else větev, neni to teda zbytečný?

balakzde commented on src/vendor_app/app/src/main/kotlin/cz/quanti/android/vendor_app/main/checkout/viewmodel/CheckoutViewModel.kt:
```diff
@@ -66,13 +73,52 @@ class CheckoutViewModel(
+            val tag = it.first
+            val userBalance = it.second
+            saveCardPurchaseToDb(
+                tag.id.asUByteArray().joinToString("") { it.toString(16).padStart(2, '0') })
```
Stejně jako u offline app, použij nějakou Util class na tohle, takovejhle kód nechceš mít na více místech

`                tag.id.asUByteArray().joinToString("") { it.toString(16).padStart(2, '0') })
`

balakzde commented on src/vendor_app/app/src/main/kotlin/cz/quanti/android/vendor_app/main/checkout/viewmodel/CheckoutViewModel.kt:
```diff
@@ -66,13 +73,52 @@ class CheckoutViewModel(
+        return Single.fromObservable(
+            nfcTagPublisher.getTagObservable().take(1).flatMapSingle { tag ->
+                nfcFacade.subtractFromBalance(tag, pin, value, currency).flatMap { userBalance ->
+                    Single.just(Pair(tag, userBalance))
```
Proč flatMap? Tady ti stačí map

balakzde commented on src/vendor_app/app/src/main/kotlin/cz/quanti/android/vendor_app/repository/booklet/impl/BookletFacadeImpl.kt:
```diff
@@ -0,0 +1,104 @@
+        return sendDeactivatedBooklets()
+    }
+
+    private fun getDataFromServer(): Completable {
```
zavádějící název. je to reload, ne get

balakzde commented on src/vendor_app/app/src/main/kotlin/cz/quanti/android/vendor_app/repository/booklet/impl/BookletRepositoryImpl.kt:
```diff
@@ -0,0 +1,114 @@
+        }
+    }
+
+    override fun getDeactivatedBookletsFromServer(): Single<Pair<Int, List<Booklet>>> {
```
Pro interface bych se raději Pair vyhnul. Vytvoř si data class, ať je hned očividný, co to vrací

balakzde commented on src/vendor_app/app/src/main/kotlin/cz/quanti/android/vendor_app/repository/card/dao/BlockedCardDao.kt:
```diff
@@ -0,0 +1,25 @@
+    fun deleteAll()
+
+    @Query("SELECT * FROM " + VendorDb.TABLE_BLOCKED_SMARTCARD + " WHERE id = :id")
+    fun getBlockedCard(id: String): Maybe<BlockedCardDbEntity>
```
https://www.sqlitetutorial.net/sqlite-exists/

balakzde commented on src/vendor_app/app/src/main/kotlin/cz/quanti/android/vendor_app/repository/login/LoginRepository.kt:
```diff
@@ -10,5 +10,5 @@ interface LoginRepository {
 
     fun login(vendor: Vendor): Single<Pair<Int, Vendor>>
 
-    fun getVendor(id: String): Single<Pair<Int, Vendor>>
```
nevracet Pair

balakzde commented on src/vendor_app/app/src/main/kotlin/cz/quanti/android/vendor_app/repository/purchase/impl/PurchaseFacadeImpl.kt:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    override fun syncWithServer(): Completable {
+        return sendPurchasesToServer()
```
Tady bacha, jak to máš naskládaný. Nejdřív pošli vouchery, pak smaž vouchery z db, pak pošli nfc a pak smaž nfc. Takhle když by failnula synchronizace u těch druhejch, tak pak ty první (vouchery) pošleš znova, ne?

balakzde commented on src/vendor_app/app/src/main/kotlin/cz/quanti/android/vendor_app/repository/purchase/impl/PurchaseFacadeImpl.kt:
```diff
@@ -0,0 +1,68 @@
+    private fun sendPurchasesToServer(): Completable {
+        return purchaseRepo.getAllPurchases().flatMapCompletable { purchases ->
+            purchaseRepo.sendVoucherPurchasesToServer(purchases)
+                .defaultIfEmpty(200)
```
Ruhle lišárnu bych asi udělal obráceně. sendVoucherPurchasesToServer by neřešila prázdnej list, to bys řešila tady

balakzde commented on src/vendor_app/app/src/main/kotlin/cz/quanti/android/vendor_app/repository/purchase/impl/PurchaseFacadeImpl.kt:
```diff
@@ -0,0 +1,68 @@
+                .defaultIfEmpty(200)
+                .flatMapCompletable { responseCode ->
+                    if (isPositiveResponseHttpCode(responseCode)) {
+                        Observable.fromIterable(purchases.filter { it.smartcard != null })
```
u nfc filtruješ, že smartcart neni null. neměla bys u voucherů filtrovat opak? Takhle posíláš i nfc platby na voucherový api, ne?

AnezkaQuanti committed:
FIX: gitignore

AnezkaQuanti committed:
FIX: Code review changes

AnezkaQuanti committed:
FIX: smartcard codes upper case should not care of locale
